augur/utils.py

```diff
@@ -143,66 +144,272 @@ def default(self, obj):
+
+    Parameters
+    ----------
+    reference : string
+        File path to GFF reference
+    feature_names : None or set or list
+        Restrict the genes we read to those in the set/list
+
+    Returns
+    -------
+    features : dict
+        keys: feature names, values: <class 'Bio.SeqFeature.SeqFeature'>
+        Note that feature names may not equivalent to GenBank feature keys
+
+    Raises
+    ------
+    AugurError
+        If the reference file contains no IDs or multiple different seqids
+        If a gene is found with the name 'nuc'
+    """
+    from BCBio import GFF
+    valid_types = ['gene', 'source', 'region']
+    limit_info = dict( gff_type = valid_types )
+    features = {}
+
+    with open(reference, encoding='utf-8') as in_handle:
+        # Note that GFF records aren't always yielded in the order one may expect,
+        # but since we raise AugurError if there are multiple this doesn't matter.
+        gff_entries = GFF.parse(in_handle, limit_info=limit_info) # generator
+        try:
+            rec = next(gff_entries)
+        except StopIteration:
+            raise AugurError(f"Reference {reference!r} contains no valid data rows. Valid GFF types (3rd column) are {', '.join(valid_types)}.")
+        # Try to read a subsequent GFF record (a separate record means rows with
+        # a different seqid) to ensure only one exists.
+        try:
+            next(gff_entries) # should raise StopIteration
+            raise AugurError(f"Reference {reference!r} contains multiple seqids (first column). Augur can only handle GFF files with a single seqid.")
+        except StopIteration:
+            pass
+
+        features['nuc'] = _read_nuc_annotation_from_gff(rec, reference)
+        features_skipped = 0
+
+        for feat in rec.features:
+            if feat.type == "gene":
+                # Check for gene names stored in qualifiers commonly used by
+                # virus-specific gene maps first (e.g., 'gene',
+                # 'gene_name'). Then, check for qualifiers used by non-viral
+                # pathogens (e.g., 'locus_tag').
+                if "gene" in feat.qualifiers:
                     fname = feat.qualifiers["gene"][0]
-                    if feature_names is None or fname in feature_names:
-                        features[fname] = feat
-            elif feat.type=='source': #read 'nuc' as well for annotations - need start/end of whole!
-                features['nuc'] = feat
+                elif "gene_name" in feat.qualifiers:
+                    fname = feat.qualifiers["gene_name"][0]
+                elif "locus_tag" in feat.qualifiers:
+                    fname = feat.qualifiers["locus_tag"][0]
+                else:
+                    features_skipped+=1
+                    fname = None
+
+                if fname == 'nuc':
+                    raise AugurError(f"Reference {reference!r} contains a gene with the name 'nuc'. This is not allowed.")
```

Comment:
Probably better for synchronous discussion, but I don't understand why we need to forbid GFF entries with a "nuc" gene name. Can't we continue to support the old style GFF formatting, so we avoid breaking existing workflows? By adding support for parsing the sequence pragma, this PR allows us to gradually update GFF files across our pathogen workflows to "do the right thing". That's a nice feature! Disallowing the existing functionality will break workflows for us and other users without warning.

If we do want to forbid the "nuc" entries, the error message should provide more details about what the user should do to fix the problem.

But, I don't think we do what to make this backward-incompatible change...

Reply:
The problem is downstream of this. A feature we give the name "nuc" will be indistinguishable from the exported JSON.annotations.nuc which must be the genome/segment/sequence length feature.

But I think i'm missing some nuance here - no genes/CDSs in Augur's test/CI have the name "nuc", and I can't think of any off the top of my head. So I don't think this'll break any existing workflows, and if it does then they were silently broken anyways!

Reply:
Ah, ok, this is my mistake. I mixed up `type=gene` and `gene=nuc` when looking at this code. We use `type=source` and `gene=nuc` in [mpox and seasonal coronavirus workflows](https://github.com/search?q=org%3Anextstrain%20gene%3Dnuc&type=code), but we don't use `type=gene` with `gene=nuc`.